examples/htmldom.py

```diff
@@ -0,0 +1,199 @@
+        self.stack = [self.root]
+        self.values: list[Any] = []
+
+    def feed(self, data: str | Thunk) -> None:
+        match data:
+            case str():
+                super().feed(data.replace("$", "$$"))
+            case getvalue, _, conv, spec:
+                self.values.append(_format_value(getvalue(), conv, spec))
+                super().feed(PLACEHOLDER)
+
+    def result(self) -> HtmlNode:
+        root = self.root
+        if len(root.children) == 1:
+            return root.children[0]
+        else:
+            return root
+
+    def handle_starttag(self, tag: str, attrs: list[tuple[str, str | None]]) -> None:
+        tag, self.values = join_with_values(tag, self.values)
+
+        attributes = {}
+        for k, v in attrs:
+            if v is not None:
+                _disallow_interpolation(k, "use attribute expansion instead")
+                attributes[k], self.values = join_with_values(v, self.values)
+            elif k == PLACEHOLDER:
+                attribute_expansion, *self.values = self.values
+                if not isinstance(attribute_expansion, dict):
+                    raise TypeError("Expected a dictionary for attribute expension")
+                attributes.update(attribute_expansion)
+            else:
+                _disallow_interpolation(k, "use attribute expansion instead")
+                attributes[k] = True
+        assert not self.values, "Did not interpolate all values"
+
+        this_node = HtmlNode(tag, attributes)
+        last_node = self.stack[-1]
+        last_node.children.append(this_node)
+        self.stack.append(this_node)
+
+    def handle_data(self, data: str) -> None:
+        interleaved_children, self.values = interleave_with_values(data, self.values)
+        assert not self.values, "Did not interpolate all values"
+
+        children = self.stack[-1].children = []
+        for child in interleaved_children:
+            match child:
+                case list() | tuple() | GeneratorType():
+                    children.extend(child)
+                case "":
+                    pass
+                case _:
+                    children.append(child)
+
+    def handle_endtag(self, tag: str) -> None:
+        self.stack.pop()
+
+
+def join_with_values(string, values) -> tuple[str, list[Any]]:
+    interleaved_values, remaining_values = interleave_with_values(string, values)
+    return "".join(map(str, interleaved_values)), remaining_values
+
+
+def interleave_with_values(string, values) -> tuple[list[str | Any], list[Any]]:
+    string_parts = string.split(PLACEHOLDER)
+    remaining_values = values[len(string_parts) - 1 :]
+
+    interleaved_values: list[str] = []
+    for s, v in zip(string_parts[:-1], values):
+        interleaved_values.append(s)
+        interleaved_values.append(v)
+    interleaved_values.append(string_parts[-1])
+
+    return interleaved_values, remaining_values
+
+
+def _format_value(value: Any, conv: str, spec: str) -> Any | str:
```

Comment:
See https://github.com/jimbaker/tagstr/blob/main/examples/taglib.py#L27 - this functionality is going to be generic, so we should have it in taglib